DynamicTablesPkg/Library/Acpi/Common/AcpiTpm2Lib/Tpm2Generator.c

```diff
@@ -0,0 +1,407 @@
+/** @file
+  TPM2 Table Generator
+
+  Copyright (c) 2022, ARM Limited. All rights reserved.
+  Copyright (c) 2023 - 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+  @par Reference(s):
+  - TCG ACPI Specification Family for TPM 1.2 and 2.0 Version 1.3 Revision 8'
+    (https://trustedcomputinggroup.org/wp-content/uploads/TCG_ACPIGeneralSpec_v1p3_r8_pub.pdf)
+
+  @par Glossary:
+  - Cm or CM   - Configuration Manager
+  - Obj or OBJ - Object
+**/
+
+#include <Library/AcpiLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Protocol/AcpiTable.h>
+
+// Module specific include files.
+#include <AcpiTableGenerator.h>
+#include <ConfigurationManagerObject.h>
+#include <ConfigurationManagerHelper.h>
+#include <Library/TableHelperLib.h>
+#include <Protocol/ConfigurationManagerProtocol.h>
+
+#include <IndustryStandard/Tpm2Acpi.h>
+
+/**
+  ARM standard TPM2 Generator
+
+  Requirements:
+    The following Configuration Manager Object(s) are used by this Generator:
+    - EArchCommonObjTpm2InterfaceInfo
+*/
+
+/**
+  This macro expands to a function that retrieves the Processor Hierarchy
+  information from the Configuration Manager.
+*/
+GET_OBJECT_LIST (
+  EObjNameSpaceArchCommon,
+  EArchCommonObjTpm2InterfaceInfo,
+  CM_ARCH_COMMON_TPM2_INTERFACE_INFO
+  );
+
+/**
+  Sanity check Start Method Specific Parameters field
+
+  @param [in]  TpmInfo            Pointer to the CM TPM2 object
+  @param [in]  AcpiTableRevision  Revision of ACPI TPM2 table
+
+  @retval EFI_SUCCESS           No failure
+  @retval EFI_INVALID_PARAMETER A parameter is invalid.
+**/
+STATIC
+EFI_STATUS
+AcpiTpm2CheckStartMethodParameters (
+  CM_ARCH_COMMON_TPM2_INTERFACE_INFO  *TpmInfo,
+  UINT8                               AcpiTableRevision
+  )
+{
+  // If LAML and LASA are present, then StartMethodParameters field would get max size
+  // regardless of StartMethod value
+  if (TpmInfo->Laml > 0) {
+    switch (AcpiTableRevision) {
+      case EFI_TPM2_ACPI_TABLE_REVISION_4:

```

Comment:
The generator should only support the revision 4 right ?
So I think this parameter should not be needed

Reply:
This is to prepare for rev 5. So I want to keep it if it's ok.

Reply:
Ok, sorry I forgot about the Rev 5.
IIUC, then the only changes required should be:
[patch.txt](https://github.com/user-attachments/files/16438416/patch.txt)

Reply:
I removed the revision check for now. It can be added later when we support revision 5.